samples/split.c

```diff
@@ -0,0 +1,182 @@
+THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE
+
+*/
+
+/* The pupose of this code is to demontrate the library apis and need proper error handling and optimization */
+
+#include <getopt.h>
+#include <unistd.h>
+#include <htslib/sam.h>
+
+/// print_usage - print the demo_usage
+/** @param fp pointer to the file / terminal to which demo_usage to be dumped
+returns nothing
+*/
+static void print_usage(FILE *fp)
+{
+    fprintf(fp, "Usage: split [-i infile] [<infile>] -o outdir\n\
+Splits the input file alignments to read1 and read2 and saves as 1.sam and 2.sam in given directory\n");
+    return;
+}
+
+/// main_demo - start of the demo
+/** @param argc - count of arguments
+ *  @param argv - pointer to array of arguments
+returns 1 on failure 0 on success
+*/
+int main(int argc, char *argv[])
+{
+    static const struct option lopts[] = {
+        {NULL, 0, NULL, 0}
+    };
+    const char *inname = NULL, *outdir = NULL;
+    char *file1 = NULL, *file2 = NULL;
+    int c = 0, ret = EXIT_FAILURE, size = 0;
+    samFile *infile = NULL, *outfile1 = NULL, *outfile2 = NULL;
+    sam_hdr_t *in_samhdr = NULL;
+    bam1_t *bamdata = NULL;
+    size_t cnt = 0;
+
+    //[-i infile] [<infile>] [-o output directory]
+    while ((c = getopt_long(argc, argv, "i:o:", lopts, NULL)) >= 0)
+    {
+        switch (c)
+        {
+        case 'i':                           //infile name
+            inname = optarg;
+            break;
+        case 'o':                           //output directory
+            outdir = optarg;
+            break;
+        case '?':                           //unknown options found!
+            print_usage(stderr);
+            goto end;
+            break;
+        }
+    }
+
+    if (argc == 1 && isatty(STDIN_FILENO)) {
+        print_usage(stdout);
+        ret = EXIT_SUCCESS;
+        goto end;
+    }
+    //infile name
+    if (!inname && ((argc - optind) == 1))
+        inname = argv[optind];
+    if (!inname)
+        inname = "-";
+    //output directory
+    if (!outdir) {
+        fprintf(stderr, "Output directory not specified\n");
+        goto end;
+    }
+    //allocate space for output
+    size = sizeof(char) * (strlen(outdir) + 10);
+    file1 = malloc(size);
+    file2 = malloc(size);
+    if (!file1 || !file2) {
+        fprintf(stderr, "Failed to set output path\n");
+        goto end;
+    }
+
+    //output file names
+    snprintf(file1, size, "%s/1.sam", outdir);
+    snprintf(file2, size, "%s/2.sam", outdir);
+    //bam data storage
+    bamdata = bam_init1();
+    if (!bamdata) {
+        fprintf(stderr, "Failed to initialize bamdata\n");
+        goto end;
+    }
+    //open input file
+    infile = sam_open(inname, "r");
+    if (!infile) {
+        fprintf(stderr, "Could not open %s\n", inname);
```

Comment:
All samtools code that invokes `sam_open()` writes its own error message itself.

Personally I think it's a bug that `sam_open` prints its own diagnostic. It is a library routine similar to `fopen` so should be quiet by default, leaving it to the invoking code to print a diagnostic — which can print a better diagnostic because it knows what the purpose of the file is (e.g. it can write “can't open reference file” or similar to distinguish further which of several input files is missing).

Some htslib library functions print their own diagnostics because it's likely that invoking code won't print its own diagnostic. However file opening functions like `sam_open` are the one case where HTSlib can depend on the invoking application code universally printing the diagnostic itself!

Reply:
Agreed on the library not reporting errors this way, but right now it does so.  Ideally it needs an hts_strerror type of interface, maybe optionally taking a file pointer (like ferror), or we split those into two APIs.    (See also comments in https://github.com/samtools/htslib/pull/1049).  A long standing wish-list!

This code serves as two purposes really.  One is standalone basic starting points that people may wish to copy and build on top of - a simple template design.  In that things like error checking and reporting are important to put in as they foster good design principles.

The second is as text quoted in a tutorial or how-to guide on using htslib.  This isn't just doxygen type manuals, which don't easily lend themselves to guidance.  For those I think we could probably do some liberal pruning of error reporting (but not detection) for the sake of brevity.  We already have some of this with liberal "..." to jump to the relevant code blocks.  It may be better to avoid putting too much in main so we can document entire functions as isolated code fragments. That also lends itself better to just returning error codes instead of printing up messages.  That depends on the application though.